a8c-files/ImageSizes.php

```diff
@@ -0,0 +1,211 @@
+			if ( isset( $_wp_additional_image_sizes[ $s ]['width'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['width'] = intval( $_wp_additional_image_sizes[ $s ]['width'] );
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['width'] = get_option( "{$s}_size_w" );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['height'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['height'] = intval( $_wp_additional_image_sizes[ $s ]['height'] );
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['height'] = get_option( "{$s}_size_h" );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['crop'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['crop'] = $_wp_additional_image_sizes[ $s ]['crop'];
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['crop'] = get_option( "{$s}_crop" );
+			}
+		}
+
+		static::$sizes = $sizes;
+
+		return $sizes;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function filtered_sizes() {
+		// Remove filter preventing the creation of advanced sizes.
+		remove_filter( 'intermediate_image_sizes_advanced', 'wpcom_intermediate_sizes' );
+
+		/** This filter is documented in wp-admin/includes/image.php */
+		$sizes = apply_filters( 'intermediate_image_sizes_advanced', static::$sizes, $this->data );
+
+		// Re-add the filter removed above.
+		add_filter( 'intermediate_image_sizes_advanced', 'wpcom_intermediate_sizes' );
+
+		return (array) $sizes;
+	}
+
+	/**
+	 * Standardises and validates the size_data array.
+	 *
+	 * @param array $size_data Size data array - at least containing height or width key. Can contain crop as well.
+	 *
+	 * @return array Array with populated width, height and crop keys; empty array if no width and height are provided.
+	 */
+	public function standardise_size_data( $size_data ) {
+		$has_at_least_width_or_height = ( isset( $size_data['width'] ) || isset( $size_data['height'] ) );
+		if ( ! $has_at_least_width_or_height ) {
+			return [];
+		}
+
+		$defaults = [
+			'width' => null,
+			'height' => null,
+			'crop' => false,
+		];
+
+		return array_merge( $defaults, $size_data );
+	}
+
+	/**
+	 * Get sizes for attachment post meta.
+	 *
+	 * @return array ImageSizes for attachment postmeta.
+	 */
+	public function generate_sizes_meta() {
+
+		$metadata = [];
+
+		foreach ( $this->filtered_sizes() as $size => $size_data ) {
+
+			$size_data = $this->standardise_size_data( $size_data );
+
+			if ( true === empty( $size_data ) ) {
+				continue;
+			}
+
+			$resized_image = $this->resize( $size_data );
+
+			if ( true === is_array( $resized_image ) ) {
+				$metadata[ $size ] = $resized_image;
+			}
+		}
+
+		return $metadata;
+	}
+
+	/**
+	 * @param array
+	 *
+	 * @return array|\WP_Error Array for usage in $metadata['sizes']; WP_Error on failure.
+	 */
+	protected function resize( $size_data ) {
+
+		$resized_image = $this->image->resize( $size_data );
```

Comment:
Unless I'm missing something, it feels like we could remove some complexity by having this method just return the data in the format of the array below, instead of changing the internal state of the image object. This would mean that we can remove the `is_resized` method and related conditionals.

Reply:
I don't think you are missing something. While reducing the Image class to a mere utilility class which would produce the array for direct usage in `$metadata['sizes']` might make the implementation less complex, it would also make it, at least from my point of view, less object oriented and less flexible.

I see the Image object in this implementation as a virtual image copy which can be modified similarly to the `WP_Image_Editor`. I have originally explored a design where there would be a class extending the `WP_Image_Editor` but performing the modifications via VIP Go File Service, but that really felt unnecessarily complex.

Unless you have strong opinions on not doing this the way it's done in the PR, I would prefer to keep it this way.

On the other hand, I have added `get_size` method, which does what you described for the `resize` one.

Please, let me know your thoughts here.

Reply:
Your reasoning makes sense, thanks!